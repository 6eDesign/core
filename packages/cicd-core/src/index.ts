import { z } from 'zod';

export { z }; // Export the zod instance

// Inline schema definitions for now
const BaseDeployableSchema = z.object({
	name: z.string(),
	__pluginType: z.string(), // Discriminator field for the plugin type
	dependencies: z.array(z.string()).optional()
});

const CommandSchema = z.object({
	type: z.string(),
	run: z.string()
});

const WorkspaceCommandsSchema = z
	.object({
		test: CommandSchema.optional(),
		build: CommandSchema.optional()
	})
	.partial();

interface ISecretProvider {
	getSecret: (name: string) => Promise<string>;
}

export interface DeployablePlugin<TInput = any> {
	input: z.Schema<TInput>;
	requiredSecrets?: string[];
	deployHandler: (
		config: TInput,
		context: {
			dryRun: boolean;
			secretProvider: ISecretProvider;
			version?: string;
			workspaceName: string;
		}
	) => Promise<any>;
}

export function createPlugin<TInput>(options: DeployablePlugin<TInput>): DeployablePlugin<TInput> {
	return options;
}

const PulumiConfigSchema = z.object({
	organization: z.string(),
	project: z.string(),
	environments: z.record(z.string(), z.string())
});

export function createCicdEngine<TPlugins extends Record<string, DeployablePlugin>>(options: {
	plugins: TPlugins;
	pulumiConfig?: z.infer<typeof PulumiConfigSchema>;
}) {
	const { plugins, pulumiConfig } = options;

	// Collect all deployable schemas for discriminated union
	const deployableSchemas = Object.entries(plugins).map(
		([pluginTypeName, plugin]) =>
			BaseDeployableSchema.extend({
				__pluginType: z.literal(pluginTypeName)
			}).merge(plugin.input as z.ZodObject<any>) // Use plugin.input directly
	);

	const cicdConfigSchema = z.object({
		deployables: z.array(
			z.discriminatedUnion(
				'__pluginType',
				deployableSchemas as [z.ZodObject<any>, ...z.ZodObject<any>[]] // Cast for discriminatedUnion
			)
		),
		commands: WorkspaceCommandsSchema.optional()
	});

	type CreateDeployableFunction = <TName extends keyof TPlugins>(
		name: TName,
		input: z.infer<TPlugins[TName]['input']> // Updated type inference
	) => ReturnType<(typeof cicdConfigSchema)['parse']>['deployables'][number];

	const createDeployable = <TName extends keyof TPlugins>(
		name: TName,
		input: z.infer<TPlugins[TName]['input']>
	) => {
		const plugin = plugins[name];
		if (!plugin) {
			throw new Error(`Unknown plugin: ${String(name)}`);
		}

		// The object is constructed here, validation happens in defineConfig
		return { __pluginType: name, ...input };
	};

	const engine = {
		defineConfig: (config: {
			deployables: Array<ReturnType<typeof createDeployable>>;
			commands?: z.infer<typeof WorkspaceCommandsSchema>;
		}) => {
			const parsedConfig = cicdConfigSchema.parse(config);
			const validatedConfig = engine.validateDependencies(parsedConfig);
			return {
				config: validatedConfig,
				engine: engine
			};
		},
		defineProjectConfig: (config: {
			environments: string[];
		}) => {
			const ProjectConfigSchema = z.object({
				environments: z.array(z.string())
			});
			return ProjectConfigSchema.parse(config);
		},
		createDeployable,
		createCommand: (run: string) => ({ type: 'default', run }),
		plugins: plugins, // Expose the plugins map
		pulumi: pulumiConfig,
		validateDependencies: (config: {
			deployables: Array<ReturnType<typeof createDeployable>>;
			commands?: z.infer<typeof WorkspaceCommandsSchema>;
		}) => {
			const deployableNames = new Set(config.deployables.map((d) => d.name));
			for (const deployable of config.deployables) {
				if (deployable.dependencies) {
					for (const dep of deployable.dependencies) {
						if (!deployableNames.has(dep)) {
							throw new Error(
								`Deployable '${deployable.name}' has an undeclared dependency: '${dep}'`
							);
						}
					}
				}
			}
			return config;
		}
	};

	return engine;
}

export type CicdEngine<TPlugins extends Record<string, DeployablePlugin>> = ReturnType<
	typeof createCicdEngine<TPlugins>
>;

export function createChildCicdEngine<TPlugins extends Record<string, DeployablePlugin>>(
	parentEngine: CicdEngine<any>,
	options: {
		plugins?: TPlugins;
		pulumiConfig?: z.infer<typeof PulumiConfigSchema>;
	}
): CicdEngine<TPlugins> {
	const mergedPlugins = { ...parentEngine.plugins, ...options.plugins };
	const mergedPulumiConfig = options.pulumiConfig || parentEngine.pulumi;

	return createCicdEngine({
		plugins: mergedPlugins,
		pulumiConfig: mergedPulumiConfig
	});
}
