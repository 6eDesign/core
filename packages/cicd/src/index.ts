import { createCicdEngine, CicdEngine } from './engine';
import { dockerImagePlugin } from './plugins/dockerImage';

const plugins = {
	dockerImage: dockerImagePlugin
};

const baseCicd: CicdEngine<typeof plugins> = createCicdEngine({
	plugins
});

export default baseCicd;

export * from './plugins/dockerImage';
export * from './engine';

// Re-export the types from the engine
export type { DeployablePlugin, CicdEngine as StackerCICDEngine } from './engine';
