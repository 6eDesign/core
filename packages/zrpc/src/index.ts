export * from './zod';
export * from './router';
export * from './service';
export * from './client';