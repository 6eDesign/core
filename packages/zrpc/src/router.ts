import { z } from './zod';
import type { Request, Response, NextFunction, Handler } from 'express';
import type { Tracer } from '@opentelemetry/api';
import { OperationObject } from 'openapi3-ts/oas30';

export type HttpMethod = 'get' | 'post' | 'put' | 'delete' | 'patch';

export interface RouteOptions<TInput, TOutput> {
	path: string;
	method?: HttpMethod;
	input: z.Schema<TInput> | z.ZodEffects<z.Schema<TInput>>;
	output: z.Schema<TOutput>;
	openapi?: OperationObject; // Add openapi property for route-level metadata
}

export const createRoute = <TInput, TOutput>({
	path,
	method = 'get',
	input,
	output,
	openapi // Destructure openapi property
}: RouteOptions<TInput, TOutput>) => ({
	path,
	method,
	input,
	output,
	openapi // Pass it through
});

const getInputGetter = <TInput, TOutput>(route: RouteOptions<TInput, TOutput>) => {
	switch (route.method) {
		case 'get':
			return (req: Request) => route.input?.parse(req.query) as TInput;
		default:
			return (req: Request) => route.input?.parse(req.body) as TInput;
	}
};

interface HttpResolverOptions<TInput, TOutput> {
	route: RouteOptions<TInput, TOutput>;
	resolver: (input: TInput) => Promise<TOutput>;
	tracer?: Tracer; // Making tracer optional for now
}

export const createHTTPResolver = <TInput, TOutput>({
	route,
	resolver
}: HttpResolverOptions<TInput, TOutput>): Handler => {
	return async (req: Request, res: Response) => {
		const inputGetter = getInputGetter(route);
		try {
			const output = await resolver(inputGetter(req));
			res.json(route.output?.parse(output));
		} catch (e: any) {
			if (e instanceof z.ZodError) {
				return res.status(400).json({ errors: e.issues });
			}
			console.error(e);
			res.status(500).json({ message: 'server error' });
		}
	};
};

export const getRouteKey = <TInput, TOutput>(route: RouteOptions<TInput, TOutput>): string => {
	return `${route.method}::${route.path}`;
};
