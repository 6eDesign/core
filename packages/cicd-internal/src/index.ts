import { createChildCicdEngine } from '@6edesign/cicd-core';
import baseCicd from '@6edesign/cicd';

export const cicd = createChildCicdEngine(baseCicd, {
  pulumiConfig: {
    organization: "stacker-org",
    project: "stacker-monorepo",
    environments: {
      dev: "dev",
      staging: "staging",
      prod: "prod",
    },
  },
});
