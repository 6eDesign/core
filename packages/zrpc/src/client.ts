import axios, { AxiosRequestConfig } from 'axios';
import { RouteOptions } from './router';

const getInputSetter = <TInput, TOutput>(route: RouteOptions<TInput, TOutput>) => {
	switch (route.method) {
		case 'get':
			return (input: TInput): Partial<AxiosRequestConfig> => ({ params: input });
		default:
			return (input: TInput): Partial<AxiosRequestConfig> => ({ data: input });
	}
};

export interface SDKOptions {
	baseUrl: string;
}

const clientMethod = <TInput, TOutput>(
	sdkOptions: SDKOptions,
	route: RouteOptions<TInput, TOutput>
) => {
	const inputSetter = getInputSetter(route);
	return (input: TInput): Promise<TOutput> =>
		axios
			.request({
				method: route.method,
				baseURL: sdkOptions.baseUrl,
				url: route.path,
				...inputSetter(input)
			})
			.then(({ data }) => route.output!.parse(data));
};

export const client = (sdkOptions: SDKOptions) => {
	return <TInput, TOutput>(
		route: RouteOptions<TInput, TOutput>
	): ((input: TInput) => Promise<TOutput>) => clientMethod(sdkOptions, route);
};

export type ClientBuilder<TRoutes> = {
	[K in keyof TRoutes]: TRoutes[K] extends RouteOptions<infer TInput, infer TOutput>
		? (input: TInput) => Promise<TOutput>
		: never;
};

export const clientFactory = <T extends Record<string, RouteOptions<any, any>>>(routes: T) => {
	return (sdkOptions: SDKOptions): ClientBuilder<T> => {
		const apiClient: any = {};
		for (const key in routes) {
			const route = routes[key];
			apiClient[key] = clientMethod(sdkOptions, route);
		}
		return apiClient as ClientBuilder<T>;
	};
};
