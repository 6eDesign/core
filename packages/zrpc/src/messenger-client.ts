import { getRouteKey, RouteOptions } from './router';
import { BaseDistributedEventBus } from '@6edesign/messenger';
import { z } from 'zod';

export interface MessageClientSDKOptions {
  eventBus: BaseDistributedEventBus;
}

const clientMethod = <TInput, TOutput>(
  channel: string,
  sdkOptions: MessageClientSDKOptions,
  route: RouteOptions<TInput, TOutput>
) => {
  return async (input: TInput): Promise<void> => {
    route.input?.parse(input);
    return sdkOptions.eventBus.send(channel, {
      key: getRouteKey(route),
      input,
    });
  };
};

export const client = (
  channel: string,
  sdkOptions: MessageClientSDKOptions
) => {
  return <TInput, TOutput>(
    route: RouteOptions<TInput, TOutput>
  ): ((input: TInput) => Promise<void>) =>
    clientMethod(channel, sdkOptions, route);
};

export const messsageClientFactory = <T extends Record<string, RouteOptions<any, any>>>(
  channel: string,
  routes: T
) => {
  return (sdkOptions: MessageClientSDKOptions) => {
    const messageClient: any = {};
    for (const key in routes) {
      const route = routes[key];
      messageClient[key] = clientMethod(channel, sdkOptions, route);
    }
    return messageClient as {
      [K in keyof T]: T[K]['input'] extends z.ZodType<any, any, any>
        ? (input: z.infer<T[K]['input']>) => Promise<void>
        : (input: void) => Promise<void>;
    };
  };
};
