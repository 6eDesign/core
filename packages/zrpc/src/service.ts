import express, { Express } from 'express';
import cors from 'cors';
import bodyParser from 'body-parser';
import responseTime from 'response-time';
import compression from 'compression';
import winston from 'winston';
import expressWinston from 'express-winston';
import { OpenAPIRegistry, OpenApiGeneratorV3 } from '@asteasolutions/zod-to-openapi';
import { createHTTPResolver, getRouteKey, RouteOptions } from './router';
import { BaseDistributedEventBus } from '@6edesign/messenger';
import { OpenAPIObject } from 'openapi3-ts/oas30';
import http from 'http';

export interface Logger {
  log: (level: string, message: string) => void;
  info: (message: string) => void;
  warn: (message: string) => void;
  error: (message: string | Error) => void;
  debug: (message: string) => void;
}

export interface ZRPCServiceOptions {
  port?: number;
  name: string;
  useCors?: boolean;
  useCompression?: boolean;
  logger?: Logger;
  eventBus?: BaseDistributedEventBus;
}

export class ZRPCService {
  public readonly app: Express;
  public readonly name: string;
  private readonly port: number;
  private readonly logger: Logger;
  private routes: RouteOptions<any, any>[] = [];
  private resolverMap: Record<string, { route: RouteOptions<any, any>; resolver: (input: any) => Promise<any> }> = {};
  private server?: http.Server;

  constructor(options: ZRPCServiceOptions) {
    this.name = options.name;
    this.port = options.port ?? 80;
    this.logger = options.logger ?? console;
    this.app = express();

    if (options.eventBus) {
      options.eventBus.listen(`${this.name}ServiceQueue`, this.handleMessage.bind(this));
    }

    if (options.useCors !== false) {
      this.app.use(cors());
    }
    if (options.useCompression !== false) {
      this.app.use(compression());
    }

    this.app.use(bodyParser.json());
    this.app.use(responseTime());

    this.app.use(expressWinston.logger({
      transports: [new winston.transports.Console()],
      format: winston.format.combine(winston.format.colorize(), winston.format.json()),
      meta: true,
      msg: 'HTTP {{req.method}} {{req.url}}',
      expressFormat: true,
      colorize: false,
    }));
  }

  public addRoute<TInput, TOutput>(
    route: RouteOptions<TInput, TOutput>,
    resolver: (input: TInput) => Promise<TOutput>
  ) {
    this.routes.push(route);
    this.resolverMap[getRouteKey(route)] = { route, resolver };

    const httpMethod = route.method ?? 'get';
    this.app[httpMethod](route.path, createHTTPResolver({ route, resolver }));
  }

  private async handleMessage({ key = '', input }: { key?: string; input?: unknown } = {}) {
    if (!key || !input) throw new Error('Invalid message format: key and input are required.');

    const { route, resolver } = this.resolverMap[key] ?? {};
    if (!resolver) {
      this.logger.error(`No resolver found for message key: ${key}`);
      throw new Error('NoResolver');
    }

    try {
      const parsedInput = route.input?.parse(input);
      await resolver(parsedInput);
    } catch (error) {
      this.logger.error(error as Error);
      throw error;
    }
  }

  public start(): Promise<void> {
    const openApiDoc = this.generateOpenAPI();
    this.app.get('/openapi.json', (req, res) => res.json(openApiDoc));
    this.app.get(['/health', '/readiness'], (req, res) => res.json({ status: 'OK' }));

    this.app.use((err: Error, req: express.Request, res: express.Response, next: express.NextFunction) => {
      this.logger.error(err.stack ?? err.message);
      res.status(500).send('Something broke!');
    });

    return new Promise((resolve) => {
      this.server = this.app.listen(this.port, () => {
        this.logger.info(`${this.name} service started on port ${this.port}`);
        resolve();
      });
    });
  }

  public stop(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (this.server) {
        this.server.close((err) => {
          if (err) {
            this.logger.error(`Error stopping service: ${err.message}`);
            return reject(err);
          }
          this.logger.info(`${this.name} service stopped.`);
          resolve();
        });
      } else {
        resolve(); // No server to stop
      }
    });
  }

  public generateOpenAPI(): OpenAPIObject {
    const registry = new OpenAPIRegistry();

    this.routes.forEach((route) => {
      registry.registerPath({
        method: route.method ?? 'get',
        path: route.path,
        ...route.openapi, // Spread the openapi property here
        request: {
          ...(route.method === 'get' && { query: route.input as any }),
          ...(route.method !== 'get' && {
            body: {
              description: '',
              content: { 'application/json': { schema: route.input as any } },
            },
          }),
        },
        responses: {
          200: {
            description: 'Success',
            content: {
              'application/json': {
                schema: route.output as any,
              },
            },
          },
        },
      });
    });

    const generator = new OpenApiGeneratorV3(registry.definitions);
    return generator.generateDocument({
      info: { title: `${this.name} API Spec`, version: '1.0' },
      servers: [{ url: `/` }],
      openapi: '3.0.0',
    });
  }
}