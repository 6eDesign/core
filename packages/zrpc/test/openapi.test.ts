import { describe, it, expect } from 'vitest';
import { z } from '../src/zod';
import { createRoute } from '../src/router';
import { ZRPCService } from '../src/service';

describe('OpenAPI Generation', () => {
  it('should generate a well-formed OpenAPI document for basic types', () => {
    const service = new ZRPCService({ name: 'BasicService', port: 0 });

    const BasicSchema = z.object({
      id: z.string().uuid(),
      name: z.string().min(3).max(255),
      age: z.number().int().positive(),
      isActive: z.boolean().default(true),
      email: z.string().email().optional(),
      createdAt: z.string().datetime(),
    });

    const getBasicRoute = createRoute({
      path: '/basic/:id',
      method: 'get',
      input: z.object({ id: z.string().uuid() }),
      output: BasicSchema,
    });

    const createBasicRoute = createRoute({
      path: '/basic',
      method: 'post',
      input: BasicSchema.omit({ id: true, createdAt: true }),
      output: BasicSchema,
    });

    service.addRoute(getBasicRoute, async (input) => ({ id: input.id, name: 'Test', age: 30, isActive: true, email: 'test@example.com', createdAt: new Date().toISOString() }));
    service.addRoute(createBasicRoute, async (input) => ({ ...input, id: 'new-id', createdAt: new Date().toISOString() }));

    const openApiDoc = service.generateOpenAPI();
    expect(openApiDoc).toMatchSnapshot();
  });

  it('should correctly represent arrays and unions', () => {
    const service = new ZRPCService({ name: 'ComplexService', port: 0 });

    const ItemSchema = z.object({
      id: z.string(),
      value: z.string(),
    });

    const StatusEnum = z.enum(['pending', 'completed', 'failed']);

    const ComplexInputSchema = z.object({
      tags: z.array(z.string()).min(1),
      data: z.array(ItemSchema).optional(),
      status: StatusEnum,
      result: z.union([z.string(), z.number()]),
    });

    const ComplexOutputSchema = z.object({
      summary: z.string(),
      details: z.union([z.array(z.string()), z.null()]),
      itemsProcessed: z.number(),
    });

    const processComplexRoute = createRoute({
      path: '/process',
      method: 'post',
      input: ComplexInputSchema,
      output: ComplexOutputSchema,
    });

    service.addRoute(processComplexRoute, async (input) => ({
      summary: `Processed ${input.tags.length} tags`,
      details: input.data ? input.data.map(item => item.value) : null,
      itemsProcessed: input.data?.length ?? 0,
    }));

    const openApiDoc = service.generateOpenAPI();
    expect(openApiDoc).toMatchSnapshot();
  });

  it('should handle optional, nullable, and default values', () => {
    const service = new ZRPCService({ name: 'OptionalService', port: 0 });

    const OptionalSchema = z.object({
      name: z.string(),
      description: z.string().optional(),
      count: z.number().default(0),
      notes: z.string().nullable(),
    });

    const getOptionalRoute = createRoute({
      path: '/optional',
      method: 'get',
      input: z.object({ name: z.string() }),
      output: OptionalSchema,
    });

    service.addRoute(getOptionalRoute, async (input) => ({
      name: input.name,
      description: 'Optional description',
      count: 5,
      notes: null,
    }));

    const openApiDoc = service.generateOpenAPI();
    expect(openApiDoc).toMatchSnapshot();
  });
});
