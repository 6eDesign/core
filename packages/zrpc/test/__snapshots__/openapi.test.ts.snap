// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OpenAPI Generation > should correctly represent arrays and unions 1`] = `
{
  "components": {
    "parameters": {},
    "schemas": {},
  },
  "info": {
    "title": "ComplexService API Spec",
    "version": "1.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/process": {
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "data": {
                    "items": {
                      "properties": {
                        "id": {
                          "type": "string",
                        },
                        "value": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "id",
                        "value",
                      ],
                      "type": "object",
                    },
                    "type": "array",
                  },
                  "result": {
                    "anyOf": [
                      {
                        "type": "string",
                      },
                      {
                        "type": "number",
                      },
                    ],
                  },
                  "status": {
                    "enum": [
                      "pending",
                      "completed",
                      "failed",
                    ],
                    "type": "string",
                  },
                  "tags": {
                    "items": {
                      "type": "string",
                    },
                    "minItems": 1,
                    "type": "array",
                  },
                },
                "required": [
                  "tags",
                  "status",
                  "result",
                ],
                "type": "object",
              },
            },
          },
          "description": "",
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "details": {
                      "anyOf": [
                        {
                          "items": {
                            "type": "string",
                          },
                          "type": "array",
                        },
                        {
                          "nullable": true,
                        },
                        {
                          "nullable": true,
                        },
                      ],
                    },
                    "itemsProcessed": {
                      "type": "number",
                    },
                    "summary": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "summary",
                    "details",
                    "itemsProcessed",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
      },
    },
  },
  "servers": [
    {
      "url": "/",
    },
  ],
}
`;

exports[`OpenAPI Generation > should generate a well-formed OpenAPI document for basic types 1`] = `
{
  "components": {
    "parameters": {},
    "schemas": {},
  },
  "info": {
    "title": "BasicService API Spec",
    "version": "1.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/basic": {
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "age": {
                    "exclusiveMinimum": true,
                    "minimum": 0,
                    "type": "integer",
                  },
                  "email": {
                    "format": "email",
                    "type": "string",
                  },
                  "isActive": {
                    "default": true,
                    "type": "boolean",
                  },
                  "name": {
                    "maxLength": 255,
                    "minLength": 3,
                    "type": "string",
                  },
                },
                "required": [
                  "name",
                  "age",
                ],
                "type": "object",
              },
            },
          },
          "description": "",
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "age": {
                      "exclusiveMinimum": true,
                      "minimum": 0,
                      "type": "integer",
                    },
                    "createdAt": {
                      "format": "date-time",
                      "type": "string",
                    },
                    "email": {
                      "format": "email",
                      "type": "string",
                    },
                    "id": {
                      "format": "uuid",
                      "type": "string",
                    },
                    "isActive": {
                      "default": true,
                      "type": "boolean",
                    },
                    "name": {
                      "maxLength": 255,
                      "minLength": 3,
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                    "age",
                    "createdAt",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
      },
    },
    "/basic/:id": {
      "get": {
        "parameters": [
          {
            "in": "query",
            "name": "id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "age": {
                      "exclusiveMinimum": true,
                      "minimum": 0,
                      "type": "integer",
                    },
                    "createdAt": {
                      "format": "date-time",
                      "type": "string",
                    },
                    "email": {
                      "format": "email",
                      "type": "string",
                    },
                    "id": {
                      "format": "uuid",
                      "type": "string",
                    },
                    "isActive": {
                      "default": true,
                      "type": "boolean",
                    },
                    "name": {
                      "maxLength": 255,
                      "minLength": 3,
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                    "age",
                    "createdAt",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
      },
    },
  },
  "servers": [
    {
      "url": "/",
    },
  ],
}
`;

exports[`OpenAPI Generation > should handle optional, nullable, and default values 1`] = `
{
  "components": {
    "parameters": {},
    "schemas": {},
  },
  "info": {
    "title": "OptionalService API Spec",
    "version": "1.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/optional": {
      "get": {
        "parameters": [
          {
            "in": "query",
            "name": "name",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "count": {
                      "default": 0,
                      "type": "number",
                    },
                    "description": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                    "notes": {
                      "nullable": true,
                      "type": "string",
                    },
                  },
                  "required": [
                    "name",
                    "notes",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
      },
    },
  },
  "servers": [
    {
      "url": "/",
    },
  ],
}
`;
